django/db/backends/__init__.py

```diff
@@ -823,6 +823,15 @@ def cache_key_culling_sql(self):
         """
         return "SELECT cache_key FROM %s ORDER BY cache_key LIMIT 1 OFFSET %%s"
 
+    def case_cast_sql(self, db_type, internal_type):
```

Comment:
I definitely do think we should at least try introducing the cast_sql method as described here. The argument types would be Field instances (and the names should likely be changed to input_field and output_field for consitency of Expression.output_field).